R/conditions.R

```diff
@@ -0,0 +1,815 @@
+#' # other handlers and to resume execution:
+#' undesirable_handler <- inplace(function(c) cat("please don't call me\n"))
+#' muffling_handler <- inplace(function(c) {
+#'   cat("muffling foo...\n")
+#'   rst_muffle(c)
+#' })
+#'
+#' with_handlers(foo = undesirable_handler,
+#'   with_handlers(foo = muffling_handler, {
+#'     cnd_signal("foo")
+#'     "return value"
+#'   }))
+#'
+#'
+#' # You can signal a critical condition with cnd_panic(). Unlike
+#' # cnd_signal() which has no side effect besides signalling the
+#' # condition, cnd_panic() makes the program terminate with an error
+#' # unless a handler can deal with the condition:
+#' \dontrun{
+#' cnd_panic(cnd)
+#' }
+#'
+#' # If you don't specify a .msg or .call, the default message/call
+#' # (supplied to cnd_new()) are displayed. Otherwise, the ones
+#' # supplied to cnd_panic() and cnd_signal() take precedence:
+#' \dontrun{
+#' critical <- cnd_new("my_error",
+#'   .msg = "default 'my_error' msg",
+#'   .call = quote(default(call))
+#' )
+#' cnd_panic(critical)
+#' cnd_panic(critical, .msg = "overridden msg")
+#'
+#' fn <- function(...) {
+#'   cnd_panic(critical, .call = TRUE)
+#' }
+#' fn(arg = foo(bar))
+#' }
+#'
+#' # Note that by default a condition signalled with cnd_panic() does
+#' # not have a muffle restart. That is because in most cases,
+#' # execution should not continue after signalling a critical
+#' # condition.
+cnd_signal <- function(.cnd, ..., .msg = NULL, .call = FALSE,
+                       .mufflable = TRUE) {
+  call <- if (.call) sys.call(-1) else NULL
+  cnd <- make_cnd(.cnd, ..., .msg = .msg, .call = call)
+  cnd_signal_(cnd, base::signalCondition, .mufflable)
+}
+#' @rdname cnd_signal
+#' @export
+cnd_panic <- function(.cnd, ..., .msg = NULL, .call = FALSE,
+                      .mufflable = FALSE) {
+  call <- if (.call) sys.call(-1) else NULL
+  cnd <- make_cnd(.cnd, ..., .msg = .msg, .call = call)
+  cnd_signal_(cnd, base::stop, .mufflable)
+}
+
+make_cnd <- function(.cnd, ..., .msg, .call) {
+  if (is_scalar_character(.cnd)) {
+    .cnd <- cnd_new(.cnd, ...)
+  } else {
+    stopifnot(is_condition(.cnd))
+  }
+
+  # Override default fields if supplied
+  .cnd$message <- .msg %||% .cnd$message %||% ""
+  .cnd$call <- .call %||% .cnd$call
+
+  .cnd
+}
+cnd_signal_ <- function(cnd, signal, mufflable) {
+  if (mufflable) {
+    class(cnd) <- c("mufflable", class(cnd))
+    withRestarts(signal(cnd), muffle = function(...) NULL)
+  } else {
+    signal(cnd)
+  }
+}
+
+#' Signal an error, warning, or message.
+#'
+#' These functions are equivalent to base functions
+#' \code{\link[base]{stop}()}, \code{\link[base]{warning}()} and
+#' \code{\link[base]{message}()}, but the \code{type} argument makes
+#' it easy to create subclassed conditions. They also don't include
+#' call information by default. This saves you from typing
+#' \code{call. = FALSE} to make error messages cleaner within package
```

Comment:
We could store the call in a `.call` field (instead of `call` which gets printed) when `FALSE`?